(DOCSP -19163): Refactor Intro for Web Devs for Readability (#1474)
mongodb/docs-realm: source/get-started/introduction-web.txt

.. default-domain:: mongodb

Building a web app usually requires sending data from a server to the browser.

> This comes with all kinds of overhead, such as authentication, and data validation. 

Reviewer: This paragraph has a few phrases which, when used together, make it feel too informal. "usually "all kinds of" "And then" here combine to make the tone feel a little off compared to the rest of our docs. We could reword slightly to improve. I think we could drop "usually" and "all kinds of" if you want to keep the "And then" - although I would probably eliminate all three.

Author: refactored a bit. kept usually for parallelism of ideas w the following paragraph 1st sentence which talks abt "common challenges".

And then the application has to be scalable and secure. 
Many libraries and frameworks exist to provide these services.
Selecting the best ones for your project requires deep understanding of the tradeoffs.

{+backend+} is a serverless application backend that streamlines solving these 

> common challenges. {+backend-short+} comes with configurable functions, integrated data access, and security rules.

Reviewer: This is a total nit, but I think we could reword here to make our point stronger. "Comes with" sounds like "When you use Realm, you also get these things on the side" - but really, we want to point out that Realm solves the problems you outlined above by "providing" or "offering" these features. This is a subtle nuance and feel free to take it or leave it, but I just wanted to point it out.
 
```suggestion
common challenges. {+backend-short+} provides configurable functions, integrated data access, and security rules.
```

.. default-domain:: mongodb

> Modern web applications are more powerful than ever before. Frontend technologies 

Author: i see your point. i just cut the paragraph and tweaked the subsequent one. 

as MongoDB Realm is a **backend** as a service, we don't need to dwell on talking about the frontend.

like React and Angular let developers create web apps that behave like native applications. 
However, as the use of web applications grows, so do the demands of maintaining them.

> Developers traditionally need to serve data to the browser from a server with a 

Author: REST is def [a standard practice](https://www.postman.com/state-of-api/api-technologies/#api-technologies), but we can sidestep mentioning it here by just removing the adjective and saying API.

RESTful API. Developers also must secure the server with features like authentication
and data validation. A wide variety of libraries and frameworks exist to address these scenarios.

> But deciding the set of solutions with the right tradeoffs for your use case can be daunting.

Reviewer 2: ```suggestion
But deciding on the set of solutions with the right tradeoffs for your use case can be daunting.
```

Author: i wasn't convinced of this change at first. but after some additional research, you're def right. adding in next batch of changes.

ref: https://english.stackexchange.com/a/62843

> But deciding the set of solutions with the right tradeoffs for your use case can be daunting.

Author: i like the change. edited slightly to make clear that it's not *just* this security stuff.

User Authentication and Management
----------------------------------
  
Secure user authentication is essential for most apps. Some apps authenticate
users with traditional email/password combinations. Others use API keys or third-party

> authentication providers like Facebook and Google. These methods can get complicated,

Reviewer: We shouldn't use "complicated" here, for similar reasons to "daunting" above. We might say something adds complexity if it adds a lot of additional code, or may be difficult to reason about logically. What if we instead reword this to something like this that provides a more specific consequence:

> Implementing these auth providers may add a lot of boilerplate code to your app. This added complexity expands the surface area for potential bugs.

Reviewer: I really do think we should remove the "complicated" here. We're establishing this problem to highlight Realm as a solution or counterpoint, and using "complicated" here implies that Realm is _not_ "complicated" when we propose it as a solution in the next paragraph. I've been through our auth provider procedures and code examples, and I don't think I'd imply that it's not "complicated" with us.

and often require a lot of boilerplate code.
 
MongoDB Realm provides built-in user management and authentication methods. These

> features encapsulate complexity, and make it easy to integrate third-party authentication

Reviewer 2: Also: could "encapsulate complexity" be simplified?
```suggestion
features encapsulate complexity and make it easy to integrate third-party authentication
```

> providers. You can enable authentication providers with a simple server-side configuration.

Reviewer: We should avoid saying "simple" here for similar reasons as above. 
```suggestion
providers. You can enable authentication providers in your server-side configuration.
```

> Then immediately log in from any client application using a Realm SDK.

Reviewer: ```suggestion
Then immediately log in from any client application using a {+backend-short+} SDK.
```

> Then immediately log in from any client application using a Realm SDK.

Reviewer 2: This doesn't make sense as a new sentence. Shouldn't it be:

>...configuration then immediately...

Author: yea you're right. i prob broke up sentence to appease the complexity analyzer tool. fixing

Schema Validation and Data Access Rules
---------------------------------------

Modern applications require that data is available using consistent types and formats.

> Data consistency guarantees that all the application's components can work with it.

Reviewer 2: Work with what?

Author: idk..poor phrasing. changing to "work together"

It's also important to make sure that any given piece of data is only accessible

> to an authorized user.  For example, you might only want to let a user have access
> to their own data.

Reviewer: We could move the "only" here to be more explicit, and we could also imply the active permission process with a slight rewording:

```suggestion
to an authorized user.  For example, you might grant a user access
to only their own data.
```

> MongoDB Realm validates data with a Realm schema you define in the `JSON schema <https://json-schema.org/>`__ standard.

Reviewer: ```suggestion
{+backend+} validates data with a {+backend-short+} schema you define in the `JSON schema <https://json-schema.org/>`__ standard.
```

> A Realm schema serves as the source of truth for defining data types in your application.

Reviewer: ```suggestion
A {+backend-short+} schema serves as the source of truth for defining data types in your application.
```

> Realm also uses the schema to map data between your application and a 

Reviewer: ```suggestion
{+backend-short+} also uses the schema to map data between your application and a 
```

`{+atlas+} <https://www.mongodb.com/realm?tck=docs_realm#atlas-form-container>`_ collection.

> You can also configure Realm to secure data with role-based data access rules.

Reviewer: ```suggestion
You can also configure {+backend-short+} to secure data with role-based data access rules.
```

> These rules dynamically determine each user's access to every document for every

Reviewer 2: Do we need the word "dynamically" here? Aside from assembling an annoying alliteration with "determine", it feels redundant when we already talk about "every request" later.

Author: i like alliteration 😄. but removing b.c unnec here

Event-Driven Serverless Functions
---------------------------------

Many applications require that some business logic runs on a server. Usually this
logic deals with sensitive user data or interacting with external services. Apps might
also need to respond in real time when something happens in the app or in an external service.

> MongoDB Realm lets you define serverless JavaScript functions called Realm Functions.

Reviewer: ```suggestion
{+backend+} lets you define serverless JavaScript functions called {+backend-short+} Functions.
```

Secure Client-Side Queries
--------------------------

> Web developers have to access their data efficiently and securely. 

Reviewer: This read a little awkwardly to me, with "developers have to" and then "you" in the next sentence. Maybe we could reword a bit so it flows more naturally; something like:

> Developers need efficient, secure access to data.

If the data is on a server, you need an API to access the data and prevent unauthorized access.

> The Realm Web SDK includes a :ref:`secure MongoDB client <intro-web-mql>` that uses

Reviewer: We could split this a little more to improve readability a little further:

```suggestion
The Realm Web SDK includes a :ref:`secure MongoDB client <intro-web-mql>` to interact with data. It uses
```

:ref:`GraphQL <intro-web-graphql>` and the :manual:`Query API </tutorial/query-documents>`
to access data in {+atlas+}. Use either or both of these APIs to work with
your application's data.

> Realm's server-side rules ensures that users only read and modify permitted document. 

Reviewer 2: ```suggestion
Realm's server-side rules ensures that users only read and modify permitted documents. 
```

Reviewer 2: I might say "documents they have access to" or something, as "permitted documents" sounds more obscure.

Rules consist of one or more user roles for each collection. A role determines if
a given user has permission to access and edit a document. Realm determines a role

> for each document in every request and enforces the role's permissions before responding.

Reviewer: Could split to improve readability:
```suggestion
for each document in every request. It enforces the role's permissions before responding.
```

Author: adding `then` to make clear it's sequential: 

```suggestion
for each document in every request. It then enforces the role's permissions before responding.
```
